crates/ty_python_semantic/src/types/narrow.rs

```diff
@@ -881,6 +881,39 @@ impl<'db, 'ast> NarrowingConstraintsBuilder<'db, 'ast> {
             .tuple_windows::<(&ruff_python_ast::Expr, &ruff_python_ast::Expr)>();
         let mut constraints = NarrowingConstraints::default();
 
+        // Narrow unions of tuples based on element checks. For example:
+        //
+        //     def _(t: tuple[int, int] | tuple[None, None]):
+        //         if t[0] is not None:
+        //             reveal_type(t)  # tuple[int, int]
+        if matches!(&**ops, [ast::CmpOp::Is | ast::CmpOp::IsNot])
```

Comment:
At some point we may want to do this for `Eq` and `NotEq` too, but that is a bit more complex to get right, can be left to a separate PR.